hashing/sha256.cpp

```diff
@@ -0,0 +1,242 @@
+            uint32_t ch = (e & f) ^ (~e & g);
+            uint32_t temp1 = h + s1 + ch + k[j] + blocks[j];
+            uint32_t s0 =
+                right_rotate(a, 2) ^ right_rotate(a, 13) ^ right_rotate(a, 22);
+            uint32_t maj = (a & b) ^ (a & c) ^ (b & c);
+            uint32_t temp2 = s0 + maj;
+
+            h = g;
+            g = f;
+            f = e;
+            e = d + temp1;
+            d = c;
+            c = b;
+            b = a;
+            a = temp1 + temp2;
+        }
+
+        // Update hash values
+        hash[0] += a;
+        hash[1] += b;
+        hash[2] += c;
+        hash[3] += d;
+        hash[4] += e;
+        hash[5] += f;
+        hash[6] += g;
+        hash[7] += h;
+    }
+
+    // Convert the hash to a hexadecimal string
+    std::string result;
+    for (size_t i = 0; i < 8; ++i) {
+        for (size_t j = 0; j < 4; ++j) {
+            uint32_t byte = (hash[i] >> (24 - j * 8)) & 0xFF;
+            std::stringstream ss;
+            ss << std::hex << std::setfill('0') << std::setw(2) << byte;
+            result += ss.str();
+        }
+    }
+
+    return result;
+}
+}  // namespace sha256
+}  // namespace hashing
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    // 1st Test
+    std::string input = "";
+    std::string expected =
+        "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
+    std::string output = hashing::sha256::sha256(input);
+    std::cout << "Input: " << input << std::endl;
+    std::cout << "Expected: " << expected << std::endl;
+    std::cout << "Output: " << output << std::endl;
+    assert(output == expected);
+    std::cout << "1st TEST PASSED\n\n";
+
+    // 2nd Test
+    input = "test";
+    expected =
+        "9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08";
+    output = hashing::sha256::sha256(input);
+    std::cout << "Input: " << input << std::endl;
+    std::cout << "Expected: " << expected << std::endl;
+    std::cout << "Output: " << output << std::endl;
+    assert(output == expected);
+    std::cout << "2nd TEST PASSED\n\n";
+
+    // 3rd Test
+    input = "Hello World";
+    expected =
+        "a591a6d40bf420404a011733cfb7b190d62c65bf0bcda32b57b277d9ad9f146e";
+    output = hashing::sha256::sha256(input);
+    std::cout << "Input: " << input << std::endl;
+    std::cout << "Expected: " << expected << std::endl;
+    std::cout << "Output: " << output << std::endl;
+    assert(output == expected);
+    std::cout << "3rd TEST PASSED\n\n";
+
+    // 4th Test
+    input = "Hello World!";
+    expected =
+        "7f83b1657ff1fc53b92dc18148a1d65dfc2d4b1fa3d677284addd200126d9069";
+    output = hashing::sha256::sha256(input);
+    std::cout << "Input: " << input << std::endl;
+    std::cout << "Expected: " << expected << std::endl;
+    std::cout << "Output: " << output << std::endl;
+    assert(output == expected);
+    std::cout << "4th TEST PASSED\n\n";
+}
```

Comment:
What value is added by printing the inputs, computed and expected hashes? What do you think about expesing these tests as:
```suggestion
static void test() {
    struct TestCase {
        const std::string input;
        const std::string expected_hash;
        TestCase(std::string input, std::string expected_hash)
            : input(std::move(input)),
              expected_hash(std::move(expected_hash)) {}
    };

    const std::vector<TestCase> test_cases{
        TestCase(
            "",
            "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"),
        TestCase(
            "test",
            "9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08"),
        TestCase(
            "Hello World",
            "a591a6d40bf420404a011733cfb7b190d62c65bf0bcda32b57b277d9ad9f146e"),
        TestCase("Hello World!",
                 "7f83b1657ff1fc53b92dc18148a1d65dfc2d4b1fa3d677284addd200126d9"
                 "069")};

    for (const auto& tc : test_cases) {
        assert(hashing::sha256::sha256(tc.input) == tc.expected_hash);
    }
}
```
The above requires `#include <vector>`.

Reply:
Sure, I'll include all these changes. Thanks for the suggestions.